### PR TITLE
En 3558/create multiple miniblocks for the same shard

### PR DESCRIPTION
**Pull request by sasurobert**
Create multiple miniblocks for the same shard. Because of the gasLimit added per miniblock, we need to create multiple miniblocks for the same destination shard. So changed the transaction coordinator, to call the createMB function for every shard multiple times, until there are no more transaction to be added into miniblocks or the time limit was reached.

### CODE REVIEW

File: process/coordinator/process.go [hunk 1]
```diff
@@ -451,29 +451,36 @@ func (tc *transactionCoordinator) CreateMbsAndProcessTransactionsFromMe(
...
-		if txSpaceRemained <= 0 {
-			break
-		}
 
-		mbSpaceRemained := int(maxMbSpaceRemained) - len(miniBlocks)
-		if mbSpaceRemained <= 0 {
-			break
-		}
+	newMBAdded := true
+	for newMBAdded == true {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/coordinator/process.go]: done

File: process/coordinator/process_test.go [hunk 2]
```diff
@@ -859,17 +867,88 @@ func TestTransactionCoordinator_CreateMbsAndProcessTransactionsFromMe(t *testing
 		return true
 	}
 
+	marshalizer := &mock.MarshalizerMock{}
+	hasher := &mock.HasherMock{}
+	for i := uint32(0); i < nrShards; i++ {
```
Comment (camilbancioiu) [inline process/coordinator/process_test.go]: Renaming `i` to `shardID` would improve the readability of the loop.

File: process/coordinator/process_test.go [hunk 3]
```diff
@@ -859,17 +867,88 @@ func TestTransactionCoordinator_CreateMbsAndProcessTransactionsFromMe(t *testing
...
+		strCache := process.ShardCacherIdentifier(0, i)
+		newTx := &transaction.Transaction{GasLimit: uint64(i)}
```
Comment (camilbancioiu) [inline process/coordinator/process_test.go]: Why is `GasLimit` equal to the shard ID?

File: process/coordinator/process_test.go [hunk 4]
```diff
@@ -859,17 +867,88 @@ func TestTransactionCoordinator_CreateMbsAndProcessTransactionsFromMe(t *testing
...
+
+	numTxsToAdd := 5
+	gasLimit := process.MaxGasLimitPerMiniBlock / uint64(numTxsToAdd)
+
+	scAddress, _ := hex.DecodeString("000000000000000000005fed9c659422cd8429ce92f8973bba2a9fb51e0eb3a1")
+	addedTxs := make([]*transaction.Transaction, 0)
+
+	allTxs := 100
+	for i := 0; i < allTxs; i++ {
+		newTx := &transaction.Transaction{GasLimit: gasLimit, GasPrice: uint64(i), RcvAddr: scAddress}
```
Comment (camilbancioiu) [inline process/coordinator/process_test.go]: Why is the `GasPrice` equal to the index of the transaction?

Comment (sasurobert) [reply process/coordinator/process_test.go]: wanted to create transactions with different hash.

File: process/coordinator/process_test.go (hunk 3 above)
Comment (sasurobert) [reply process/coordinator/process_test.go]: create transaction with different hash

File: process/block/preprocess/basePreProcess.go [hunk 5]
```diff
@@ -223,3 +223,15 @@ func (bpp *basePreProcess) computeExistingAndMissing(
 
 	return missingTxsForShard
 }
+
+func (bpp *basePreProcess) isTxAlreadyProcessed(txHash []byte, forBlock *txsForBlock) bool {
+	forBlock.mutTxsForBlock.RLock()
+	txInfo := forBlock.txHashAndInfo[string(txHash)]
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/preprocess/basePreProcess.go]: done

File: process/coordinator/process_test.go (hunk 2 above)
Comment (sasurobert) [reply process/coordinator/process_test.go]: done

File: process/coordinator/process.go [hunk 6]
```diff
@@ -451,29 +451,36 @@ func (tc *transactionCoordinator) CreateMbsAndProcessTransactionsFromMe(
...
-		miniBlock, err := txPreProc.CreateAndProcessMiniBlock(
-			tc.shardCoordinator.SelfId(),
-			uint32(i),
-			txSpaceRemained,
-			haveTime,
-			round)
-		if err != nil {
-			continue
-		}
+		for i := 0; i < int(tc.shardCoordinator.NumberOfShards()); i++ {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/coordinator/process.go]: done

File: process/coordinator/process_test.go (hunk 3 above)
Comment (camilbancioiu) [reply process/coordinator/process_test.go]: Ok, no problem. Curious though, why not use the transaction nonce?

File: process/coordinator/process_test.go (hunk 4 above)
Comment (camilbancioiu) [reply process/coordinator/process_test.go]: Ok, no problem. Curious though, why not use the transaction nonce?

**Pull request merged**